File: dodam-system-domain/dodam-domain-rds/src/main/java/b1nd/dodam/domain/rds/club/service/ClubMemberService.java
package b1nd.dodam.domain.rds.club.service;

import b1nd.dodam.domain.rds.club.entity.Club;
import b1nd.dodam.domain.rds.club.entity.ClubMember;
import b1nd.dodam.domain.rds.club.enumeration.ClubPermission;
import b1nd.dodam.domain.rds.club.enumeration.ClubStatus;
import b1nd.dodam.domain.rds.club.enumeration.ClubType;
import b1nd.dodam.domain.rds.club.exception.*;
import b1nd.dodam.domain.rds.club.repository.ClubMemberRepository;
import b1nd.dodam.domain.rds.club.repository.ClubRepository;
import b1nd.dodam.domain.rds.member.entity.Member;
import b1nd.dodam.domain.rds.member.entity.Student;
import b1nd.dodam.domain.rds.member.enumeration.ActiveStatus;
import b1nd.dodam.domain.rds.member.repository.StudentRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
@RequiredArgsConstructor
public class ClubMemberService {
    private final ClubRepository clubRepository;
    private final ClubMemberRepository clubMemberRepository;
    private final StudentRepository studentRepository;

    public void saveClubMembers(List<ClubMember> clubMembers) {
        clubMemberRepository.saveAll(clubMembers);
    }

    public void saveAndValidateClubMembers(List<ClubMember> clubMember) {
        clubMember.forEach(c -> validateByClubAndStudent(c.getClub(), c.getStudent()));
        clubMemberRepository.saveAll(clubMember);
    }

    public void setClubMemberStatus(Long clubMemberId, Member member, ClubStatus clubStatus) {
        Student student = studentRepository.getByMember(member);
        ClubMember clubMember = clubMemberRepository.getByIdAndStudent(clubMemberId, student);
        if (clubStatus == ClubStatus.ALLOWED && clubMember.getClub().getType() == ClubType.CREATIVE_ACTIVITY_CLUB) {
            rejectActivityClubMember(student);
        }
        clubMember.modifyStatus(clubStatus);
        clubMemberRepository.save(clubMember);
    }

    public List<Student> getSecondGradeStudent(Member member) {
        return clubMemberRepository.findSecondGradeStudentsNotInClubMember(ActiveStatus.ACTIVE, member);
    }

    public List<Student> getAllGradeStudent(Member member) {
        return studentRepository.findAllByMember_StatusAndMemberNot(ActiveStatus.ACTIVE, member);
    }


    public List<ClubMember> findUserAllowedClub(Member member) {
        return clubMemberRepository.findByStudentAndClubStatusAndClub_State(studentRepository.getByMember(member), ClubStatus.ALLOWED, ClubStatus.ALLOWED);
    }

    public void setStatusStudentClub(int studentId, Long clubId, ClubStatus clubStatus) {
        ClubMember clubMember = clubMemberRepository.getPendingClubMemberWithRelations(studentId, clubId, ClubStatus.PENDING);
        if (clubStatus == ClubStatus.ALLOWED && clubMember.getStudent().getGrade() == 1) {
            long count = clubMemberRepository.countByClubIdAndClubStatusAndStudentGrade(clubId, ClubStatus.ALLOWED, 1);
            if (count >= clubMember.getClub().getMaxMemberCount()) {
                throw new OverflowMemberSizeException();
            }
        }
        clubMember.modifyStatus(clubStatus);
        clubMemberRepository.save(clubMember);
    }

    public List<ClubMember> getAllLeader() {
        return clubMemberRepository.findAllByPermission(ClubPermission.CLUB_LEADER);
    }

    public Club findClubIfNotClubMember(Long clubId, ClubStatus state, Student student, ClubStatus status) {
        return clubMemberRepository.findClubIfNotMember(clubId, state, student, status).orElseThrow(ClubJoinedException::new);
    }

    public List<ClubMember> findAllCreativeClubByStudent(Student student) {
        List<ClubMember> clubMembers = clubMemberRepository.findByStudentAndClubStatusInAndClub_TypeAndClub_State(student, List.of(ClubStatus.PENDING), ClubType.CREATIVE_ACTIVITY_CLUB, ClubStatus.ALLOWED);
        if(clubMembers.isEmpty()) {
            throw new ClubNotFoundException();
        }
        return clubMembers;
    }

    public void updateStatus(List<ClubMember> members, ClubStatus status) {
        members.forEach(member -> member.modifyStatus(status));
        List<Student> students = members.stream().map(ClubMember::getStudent).toList();
        List<ClubMember> selectedClubMembers = clubMemberRepository.findByStudentInAndClubStatusNotAndClub_Type(students, status, ClubType.CREATIVE_ACTIVITY_CLUB);
        selectedClubMembers.forEach(clubMember -> clubMember.modifyStatus(ClubStatus.REJECTED));
    }

    public List<Club> getStudentClubStatus(Student student) {
        return clubMemberRepository.findByStudentAndPermissionAndClub_StateNot(student, ClubPermission.CLUB_LEADER, ClubStatus.DELETED)
                .stream()
                .map(ClubMember::getClub)
                .collect(Collectors.toList());
    }

    public List<ClubMember> getJoinRequests(Member member, ClubStatus status) {
        Student student = studentRepository.getByMember(member);
        return clubMemberRepository.findByStudentAndClubStatus(student, status);
    }

    public List<ClubMember> getJoinRequests(Member member) {
        Student student = studentRepository.getByMember(member);
        return clubMemberRepository.findAllByStudent(student);
    }

    public ClubMember getClubLeader(Long clubId) {
        return clubMemberRepository.getByClubAndPermissionAndStatus(clubRepository.getByClubId(clubId), ClubPermission.CLUB_LEADER, ClubStatus.ALLOWED);
    }

    public void validateActiveClubMemberSize(Club club, Member leader) {
        validateByClubLeader(club, leader);
        if (clubMemberRepository.findAllByClubAndClubStatus(club, ClubStatus.ALLOWED).size() < 5) {
            throw new InsufficientClubMembersException();
        }
    }

    public List<ClubMember> getStatusClubMembers(Long clubId,  ClubStatus clubStatus) {
        return clubMemberRepository.findAllByClubAndClubStatus(clubRepository.getByClubId(clubId), clubStatus);
    }

    public List<ClubMember> getAllClubMembers(Long clubId) {
        Club club = clubRepository.getByClubId(clubId);
        return clubMemberRepository.findAllByClubAndPermission(club, ClubPermission.CLUB_MEMBER);
    }

    public void setDeleteAllClubMembers(Club club) {
        List<ClubMember> clubMembers = clubMemberRepository.findByClub(club);
        clubMembers.forEach(cm -> cm.modifyStatus(ClubStatus.DELETED));
        clubMemberRepository.saveAll(clubMembers);
    }

    public void setDeleteClubMembers(Club club) {
        List<ClubMember> clubMembers = clubMemberRepository.findByClubAndClubStatusNot(club, ClubStatus.ALLOWED);
        clubMembers.forEach(cm -> cm.modifyStatus(ClubStatus.DELETED));
        clubMemberRepository.saveAll(clubMembers);
    }

    public void validateAndRejectLeader(Club club, Student leader, List<Student> students) {
        if (club.getType() == ClubType.CREATIVE_ACTIVITY_CLUB) {
            rejectActivityClubMember(leader);
        }
        validateLeaderInList(leader, students);
        validateByLeaderAndClubMemberDuplicated(leader, students, club);
    }

    public void validateByClubLeader(Club club, Member member) {
        Student leader = studentRepository.getByMember(member);
        if(!clubMemberRepository.existsByClubAndStudentAndPermission(club, leader, ClubPermission.CLUB_LEADER)) {
            throw new ClubPermissionDeniedException();
        }
    }

    public void validateByClubAndStudent(Club clubId, Student student) {
        if (!(clubMemberRepository.findByClubAndStudentAndClubStatusNot(clubId, student, ClubStatus.DELETED) == null)) {
            throw new ClubJoinedException();
        }
    }

    public void validateFirstGrade(Student student) {
        if (student.getGrade() != 1) {
            throw new ClubApplicationNotAllowedException();
        }
    }

    public void validateNoActiveCreativeClub(Student student) {
        boolean hasActive = !clubMemberRepository.findByStudentAndClubStatusInAndClub_TypeAndClub_State(
                student, List.of(ClubStatus.ALLOWED, ClubStatus.PENDING),
                ClubType.CREATIVE_ACTIVITY_CLUB, ClubStatus.ALLOWED
        ).isEmpty();
        if (hasActive) {
            throw new ClubJoinedException();
        }
    }

    public boolean isClubLeader(Long clubId,  Member member) {
        Student leader = studentRepository.findByMember(member).orElse(null);
        if (leader == null) {
            return false;
        }
        Club club = clubRepository.getByClubId(clubId);
        ClubMember clubMember = clubMemberRepository.findByClubAndStudentAndPermission(club, leader, ClubPermission.CLUB_LEADER);
        return clubMember != null;
    }

    public long countFirstGradeAllowedMembers(Club club) {
        return clubMemberRepository.countByClubIdAndClubStatusAndStudentGrade(club.getId(), ClubStatus.ALLOWED, 1);
    }

    public void autoApproveClubsWithinLimit() {
        List<ClubMember> pendingFirstGrade = clubMemberRepository.findPendingFirstGradeByClubType(
                ClubStatus.PENDING, 1, ClubType.CREATIVE_ACTIVITY_CLUB, ClubStatus.ALLOWED
        );

        Map<Club, List<ClubMember>> byClub = pendingFirstGrade.stream()
                .collect(Collectors.groupingBy(ClubMember::getClub));

        byClub.forEach((club, members) -> {
            if (members.size() <= club.getMaxMemberCount()) {
                members.forEach(m -> m.modifyStatus(ClubStatus.ALLOWED));
                clubMemberRepository.saveAll(members);
            }
        });
    }

    private void rejectActivityClubMember(Student student) {
        List<ClubMember> clubMembers = clubMemberRepository.findAllByStudentAndPermissionAndClub_Type(student, ClubPermission.CLUB_MEMBER, ClubType.CREATIVE_ACTIVITY_CLUB);
        clubMembers.forEach(m -> m.modifyStatus(ClubStatus.REJECTED));
        clubMemberRepository.saveAll(clubMembers);
    }

    private void validateLeaderInList(Student leader, List<Student> students) {
        if (students.stream().anyMatch(s -> s.getId() == leader.getId())) {
            throw new InvalidClubMemberInviteException();
        }
    }

    private void validateByLeaderAndClubMemberDuplicated(Student leader, List<Student> students, Club club) {
        if (club.getType() == ClubType.SELF_DIRECT_ACTIVITY_CLUB) {
            return;
        }
        if (!clubMemberRepository.findByStudentInAndClubStatusAndClub_TypeAndClub_StateNot(
                Stream.concat(Stream.of(leader), students.stream()).toList(),
                ClubStatus.ALLOWED,
                ClubType.CREATIVE_ACTIVITY_CLUB,
                ClubStatus.DELETED).isEmpty()
        ) {
            throw new InvalidClubMemberInviteException();
        }
    }
}

File: services/service-club/src/main/kotlin/com/b1nd/dodamdodam/club/domain/club/service/ClubMemberService.kt
package com.b1nd.dodamdodam.club.domain.club.service

import com.b1nd.dodamdodam.club.domain.club.entity.ClubEntity
import com.b1nd.dodamdodam.club.domain.club.entity.ClubMemberEntity
import com.b1nd.dodamdodam.club.domain.club.enumeration.ClubPermission
import com.b1nd.dodamdodam.club.domain.club.enumeration.ClubPriority
import com.b1nd.dodamdodam.club.domain.club.enumeration.ClubStatus
import com.b1nd.dodamdodam.club.domain.club.enumeration.ClubType
import com.b1nd.dodamdodam.club.domain.club.exception.ClubJoinedException
import com.b1nd.dodamdodam.club.domain.club.exception.ClubMemberNotFoundException
import com.b1nd.dodamdodam.club.domain.club.exception.ClubNotFoundException
import com.b1nd.dodamdodam.club.domain.club.exception.ClubPermissionDeniedException
import com.b1nd.dodamdodam.club.domain.club.exception.InsufficientClubMembersException
import com.b1nd.dodamdodam.club.domain.club.exception.InvalidClubMemberInviteException
import com.b1nd.dodamdodam.club.domain.club.repository.ClubMemberRepository
import com.b1nd.dodamdodam.club.domain.club.repository.ClubRepository
import org.springframework.stereotype.Service

@Service
class ClubMemberService(
    private val clubMemberRepository: ClubMemberRepository,
    private val clubRepository: ClubRepository,
) {
    fun createClubMembers(members: List<ClubMemberEntity>) {
        clubMemberRepository.saveAll(members)
    }

    fun createAndValidateClubMembers(members: List<ClubMemberEntity>) {
        members.forEach { member ->
            validateByClubAndStudent(member.club, member.studentId)
        }
        clubMemberRepository.saveAll(members)
    }

    fun setClubMemberStatus(clubMemberId: Long, studentId: Long, clubStatus: ClubStatus) {
        val clubMember = clubMemberRepository.findByIdAndStudentId(clubMemberId, studentId)
            ?: throw ClubMemberNotFoundException()

        if (clubStatus == ClubStatus.REJECTED && clubMember.club.type == ClubType.CREATIVE_ACTIVITY_CLUB) {
            rejectOtherCreativeClubs(clubMember.studentId)
        }

        clubMember.modifyStatus(clubStatus)
        clubMemberRepository.save(clubMember)
    }

    fun setStatusStudentClub(studentId: Long, clubId: Long, status: ClubStatus) {
        val club = clubRepository.findById(clubId)
            .orElseThrow { ClubNotFoundException() }
        val clubMember = clubMemberRepository.findByClubAndStudentIdAndClubStatusNot(club, studentId, ClubStatus.DELETED)
            ?: throw ClubMemberNotFoundException()
        clubMember.modifyStatus(status)
        clubMemberRepository.save(clubMember)
    }

    fun getAllLeaders(): List<ClubMemberEntity> =
        clubMemberRepository.findAllByPermission(ClubPermission.CLUB_LEADER)

    fun findClubIfNotClubMember(clubId: Long, state: ClubStatus, studentId: Long, excludeStatus: ClubStatus): ClubEntity {
        val club = clubRepository.findByIdAndStateNot(clubId, ClubStatus.DELETED)
            ?: throw ClubNotFoundException()
        val existingMember = clubMemberRepository.findByClubAndStudentIdAndClubStatusNot(club, studentId, excludeStatus)
        if (existingMember != null) {
            throw ClubJoinedException()
        }
        return club
    }

    fun findAllCreativeClubByStudent(studentId: Long): List<ClubMemberEntity> =
        clubMemberRepository.findByStudentIdAndClubStatusInAndClub_TypeAndClub_State(
            studentId,
            listOf(ClubStatus.PENDING, ClubStatus.ALLOWED),
            ClubType.CREATIVE_ACTIVITY_CLUB,
            ClubStatus.ALLOWED,
        )

    fun getPendingAndAllowedMembersByClubs(clubs: List<ClubEntity>): List<ClubMemberEntity> =
        clubMemberRepository.findByClubInAndClubStatusNotIn(
            clubs,
            ClubStatus.getNotAllowedStatuses(),
        )

    fun updateStatus(members: List<ClubMemberEntity>, status: ClubStatus) {
        members.forEach { member ->
            member.modifyStatus(status)

            if (status == ClubStatus.ALLOWED && member.club.type == ClubType.CREATIVE_ACTIVITY_CLUB) {
                rejectOtherCreativeClubs(member.studentId)
            }
        }
        clubMemberRepository.saveAll(members)
    }

    fun getStudentClubStatus(studentId: Long): List<ClubEntity> {
        val leaderMemberships = clubMemberRepository.findByStudentIdAndPermissionAndClub_StateNot(
            studentId,
            ClubPermission.CLUB_LEADER,
            ClubStatus.DELETED,
        )
        return leaderMemberships.map { it.club }
    }

    fun getJoinRequests(studentId: Long, status: ClubStatus): List<ClubMemberEntity> =
        clubMemberRepository.findByStudentIdAndClubStatus(studentId, status)

    fun getClubLeader(clubId: Long): ClubMemberEntity {
        val club = clubRepository.findById(clubId)
            .orElseThrow { ClubNotFoundException() }
        return clubMemberRepository.findByClubAndPermissionAndClubStatus(
            club,
            ClubPermission.CLUB_LEADER,
            ClubStatus.ALLOWED,
        ) ?: throw ClubMemberNotFoundException()
    }

    fun validateActiveClubMemberSize(club: ClubEntity, leaderStudentId: Long) {
        val leader = clubMemberRepository.findByClubAndStudentIdAndPermission(
            club,
            leaderStudentId,
            ClubPermission.CLUB_LEADER,
        ) ?: throw ClubPermissionDeniedException()

        val activeMembers = clubMemberRepository.findAllByClubAndClubStatus(club, ClubStatus.ALLOWED)
        if (activeMembers.size < 5) {
            throw InsufficientClubMembersException()
        }
    }

    fun getStatusClubMembers(clubId: Long, status: ClubStatus): List<ClubMemberEntity> {
        val club = clubRepository.findById(clubId)
            .orElseThrow { ClubNotFoundException() }
        return clubMemberRepository.findByClubAndClubStatus(club, status)
    }

    fun getAllClubMembers(clubId: Long): List<ClubMemberEntity> {
        val club = clubRepository.findById(clubId)
            .orElseThrow { ClubNotFoundException() }
        return clubMemberRepository.findByClub(club)
    }

    fun setDeleteAllClubMembers(club: ClubEntity) {
        val members = clubMemberRepository.findByClub(club)
        members.forEach { it.modifyStatus(ClubStatus.DELETED) }
        clubMemberRepository.saveAll(members)
    }

    fun setDeleteClubMembers(club: ClubEntity) {
        val members = clubMemberRepository.findByClubAndClubStatusNot(club, ClubStatus.ALLOWED)
        members.forEach { it.modifyStatus(ClubStatus.DELETED) }
        clubMemberRepository.saveAll(members)
    }

    fun validateAndRejectLeader(club: ClubEntity, leaderStudentId: Long, studentIds: List<Long>) {
        if (club.type == ClubType.CREATIVE_ACTIVITY_CLUB) {
            studentIds.forEach { studentId ->
                val isJoined = isCreativeClubJoined(studentId)
                if (isJoined) {
                    throw InvalidClubMemberInviteException()
                }
            }
        }

        if (studentIds.contains(leaderStudentId)) {
            throw InvalidClubMemberInviteException()
        }

        if (studentIds.size != studentIds.distinct().size) {
            throw InvalidClubMemberInviteException()
        }
    }

    fun validateByClubLeader(club: ClubEntity, studentId: Long) {
        val isLeader = clubMemberRepository.existsByClubAndStudentIdAndPermission(
            club,
            studentId,
            ClubPermission.CLUB_LEADER,
        )
        if (!isLeader) {
            throw ClubPermissionDeniedException()
        }
    }

    fun validateByClubAndStudent(club: ClubEntity, studentId: Long) {
        val existingMember = clubMemberRepository.findByClubAndStudentIdAndClubStatusNot(
            club,
            studentId,
            ClubStatus.DELETED,
        )
        if (existingMember != null) {
            throw ClubJoinedException()
        }
    }

    fun isCreativeClubJoined(studentId: Long): Boolean {
        val memberships = clubMemberRepository.findByStudentIdAndClubStatusInAndClub_Type(
            studentId,
            listOf(ClubStatus.PENDING, ClubStatus.ALLOWED),
            ClubType.CREATIVE_ACTIVITY_CLUB,
        )
        return memberships.isNotEmpty()
    }

    fun isClubLeader(clubId: Long, studentId: Long): Boolean {
        val club = clubRepository.findById(clubId).orElse(null) ?: return false
        return clubMemberRepository.existsByClubAndStudentIdAndPermission(
            club,
            studentId,
            ClubPermission.CLUB_LEADER,
        )
    }

    fun shuffleClubMemberMap(
        maxCount: Int,
        clubMemberMap: Map<ClubEntity, List<ClubMemberEntity>>,
        priority: ClubPriority,
    ): List<ClubMemberEntity> {
        val selectedMembers = mutableListOf<ClubMemberEntity>()

        clubMemberMap.forEach { (_, members) ->
            val priorityMembers = members.filter { it.priority == priority }
            val shuffled = priorityMembers.shuffled()
            val selected = shuffled.take(maxCount)
            selectedMembers.addAll(selected)
        }

        return selectedMembers
    }

    private fun rejectOtherCreativeClubs(studentId: Long) {
        val otherMemberships = clubMemberRepository.findByStudentIdAndClubStatusAndClub_State(
            studentId,
            ClubStatus.PENDING,
            ClubStatus.ALLOWED,
        ).filter { it.club.type == ClubType.CREATIVE_ACTIVITY_CLUB }

        otherMemberships.forEach { it.modifyStatus(ClubStatus.REJECTED) }
        if (otherMemberships.isNotEmpty()) {
            clubMemberRepository.saveAll(otherMemberships)
        }
    }
}
